cmd/observer/node_builder/observer_builder.go

```diff
@@ -0,0 +1,1018 @@
+			FetchTimeout:       edrequester.DefaultFetchTimeout,
+			RetryDelay:         edrequester.DefaultRetryDelay,
+			MaxRetryDelay:      edrequester.DefaultMaxRetryDelay,
+		},
+		apiTimeout:             3 * time.Second,
+		upstreamNodeAddresses:  []string{},
+		upstreamNodePublicKeys: []string{},
+	}
+}
+
+// ObserverServiceBuilder provides the common functionality needed to bootstrap a Flow observer service
+// It is composed of the FlowNodeBuilder, the ObserverServiceConfig and contains all the components and modules needed for the observers
+type ObserverServiceBuilder struct {
+	*cmd.FlowNodeBuilder
+	*ObserverServiceConfig
+
+	// components
+	LibP2PNode              *p2p.Node
+	FollowerState           protocol.MutableState
+	SyncCore                *synchronization.Core
+	RpcEng                  *rpc.Engine
+	FinalizationDistributor *pubsub.FinalizationDistributor
+	FinalizedHeader         *synceng.FinalizedHeaderCache
+	Committee               hotstuff.Committee
+	Finalized               *flow.Header
+	Pending                 []*flow.Header
+	FollowerCore            module.HotStuffFollower
+	ExecutionDataService    state_synchronization.ExecutionDataService
+	ExecutionDataRequester  state_synchronization.ExecutionDataRequester // for the observer, the sync engine participants provider is the libp2p peer store which is not
+	// available until after the network has started. Hence, a factory function that needs to be called just before
+	// creating the sync engine
+	SyncEngineParticipantsProviderFactory func() id.IdentifierProvider
+
+	// engines
+	FollowerEng *followereng.Engine
+	SyncEng     *synceng.Engine
+
+	// Public network
+	peerID peer.ID
+}
+
+// deriveBootstrapPeerIdentities derives the Flow Identity of the bootstrap peers from the parameters.
+// These are the identities of the observers also acting as the DHT bootstrap server
+func (builder *ObserverServiceBuilder) deriveBootstrapPeerIdentities() error {
+	// if bootstrap identities already provided (as part of alternate initialization as a library the skip reading command
+	// line params)
+	if builder.bootstrapIdentities != nil {
+		return nil
+	}
+
+	ids, err := BootstrapIdentities(builder.bootstrapNodeAddresses, builder.bootstrapNodePublicKeys)
+	if err != nil {
+		return fmt.Errorf("failed to derive bootstrap peer identities: %w", err)
+	}
+
+	builder.bootstrapIdentities = ids
+
+	return nil
+}
+
+// deriveBootstrapPeerIdentities derives the Flow Identity of the bootstrap peers from the parameters.
+// These are the identities of the observers also acting as the DHT bootstrap server
+func (builder *ObserverServiceBuilder) deriveUpstreamPeerIdentities() error {
+	// if bootstrap identities already provided (as part of alternate initialization as a library the skip reading command
+	// line params)
+	if builder.upstreamIdentities != nil {
+		return nil
+	}
+
+	ids, err := BootstrapIdentities(builder.upstreamNodeAddresses, builder.upstreamNodePublicKeys)
```

Comment:
as mentioned in a previous comment, I don't think it's safe to use `BootstrapIdentities` here

Reply:
Addressed: https://github.com/GetElastech/flow-go/commit/90f72719c012a44cde6c0261d2b3abe2c259b8cf